core-active-state.html

```diff
@@ -0,0 +1,175 @@
+<script>
+  Polymer({
+    is: 'core-active-state',
+
+    properties: {
+
+      /**
+       * If true, the element is currently active either because the
+       * user is touching it, or the button is a toggle
+       * and is currently in the active state.
+       *
+       * @attribute active
+       * @type boolean
+       * @default false
+       */
+      active: {
+        type: Boolean,
+        notify: true,
+        readOnly: true,
+        observer: 'activeChanged',
+        value: false
+      },
+
+      /**
+       * If true, the user is currently holding down the button.
+       *
+       * @attribute pressed
+       * @type boolean
+       * @default false
+       */
+      pressed: {
+        type: Boolean,
+        value: function() {
+          this._setPressed(this.pressed || false);
+        },
+        readOnly: true,
+        observer: 'pressedChanged'
+      },
+
+      /**
+       * If true, the button toggles the active state with each tap.
+       * Otherwise, the button becomes active when the user is holding
+       * it down.
+       *
+       * @attribute toggle
+       * @type boolean
+       * @default false
+       */
+      toggle: {
+        type: Boolean,
+        value: false
+      },
+
+      /**
+       * If true, the user cannot interact with this element.
+       *
+       * @attribute disabled
+       * @type boolean
+       * @default false
+       */
+      disabled: {
+        type: Boolean,
+        observer: 'disabledChanged',
+        value: false
+      }
+    },
+
+    get target() {
+      return this.host || this.parentNode;
+    },
+
+    configure: function() {
+      return {
+        boundDownHandler: this.onTargetDown.bind(this),
+        boundUpHandler: this.onTargetUp.bind(this)
+      };
+    },
+
+    ready: function() {
+      this.afterReady = true;
+      // TODO(cdata): We this as a workaround for polymer/polymer#1285
+      if (this.isAttached) {
+        this.attached();
+      }
+    },
+
+    attached: function() {
+      if (!this.afterReady) {
+        return;
+      }
+
+      this.target.addEventListener('mousedown', this.boundDownHandler);
+      this.target.addEventListener('mouseup', this.boundUpHandler);
+      this.target.addEventListener('keydown', this.boundDownHandler);
+      this.target.addEventListener('keyup', this.boundUpHandler);
+    },
+
+    detached: function() {
+      this.target.removeEventListener('mousedown', this.boundDownHandler);
+      this.target.removeEventListener('mouseup', this.boundUpHandler);
+      this.target.removeEventListener('keydown', this.boundDownHandler);
+      this.target.removeEventListener('keyup', this.boundUpHandler);
+    },
+
+    onTargetDown: function(event) {
+      if (event.type === 'keydown' &&
+          event.keyCode !== 32 &&
+          event.keyCode !== 13) {
+        return;
+      }
+
+      if (event.keyCode === 13) {
+        window.setTimeout(function() {
```

Comment:
> Ah! Ok, so you're running into the joy that are tasks and microtasks:

Thanks for the explanation :) I had a good talk with Steve about it this morning, and now things make a lot more sense to me.

> subsequent calls are clamped to a 4ms delay.

I had no idea that this was the case. That's very obscure (and useful) information to have.

> However, if you want the click event to fire after keyup, you should be performing this logic in onTargetUp, not onTargetDown!

Not true! This could be made more clear, but in the platform button a click is triggered on `keydown` when the key being pressed is `enter`. The click is triggered on `keyup` when the key being pressed is `space`. Fun, right?